crates/matrix-sdk-sqlite/src/event_cache_store.rs

```diff
@@ -0,0 +1,385 @@
+    }
+
+    async fn remove_media(&self, uri: Key, format: Key) -> Result<()> {
+        self.execute("DELETE FROM media WHERE uri = ? AND format = ?", (uri, format)).await?;
+        Ok(())
+    }
+
+    async fn remove_uri_medias(&self, uri: Key) -> Result<()> {
+        self.execute("DELETE FROM media WHERE uri = ?", (uri,)).await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl SqliteObjectEventCacheStoreExt for deadpool_sqlite::Object {}
+
+#[async_trait]
+impl EventCacheStore for SqliteEventCacheStore {
+    type Error = Error;
+
+    async fn add_media_content(&self, request: &MediaRequest, content: Vec<u8>) -> Result<()> {
+        let uri = self.encode_key(keys::MEDIA, request.source.unique_key());
+        let format = self.encode_key(keys::MEDIA, request.format.unique_key());
+        let data = self.encode_value(content)?;
+        self.acquire().await?.set_media(uri, format, data).await
+    }
+
+    async fn get_media_content(&self, request: &MediaRequest) -> Result<Option<Vec<u8>>> {
+        let uri = self.encode_key(keys::MEDIA, request.source.unique_key());
+        let format = self.encode_key(keys::MEDIA, request.format.unique_key());
+        let data = self.acquire().await?.get_media(uri, format).await?;
+        data.map(|v| self.decode_value(&v).map(Into::into)).transpose()
+    }
+
+    async fn remove_media_content(&self, request: &MediaRequest) -> Result<()> {
+        let uri = self.encode_key(keys::MEDIA, request.source.unique_key());
+        let format = self.encode_key(keys::MEDIA, request.format.unique_key());
+        self.acquire().await?.remove_media(uri, format).await
+    }
+
+    async fn remove_media_content_for_uri(&self, uri: &ruma::MxcUri) -> Result<()> {
+        let uri = self.encode_key(keys::MEDIA, uri);
+        self.acquire().await?.remove_uri_medias(uri).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        sync::atomic::{AtomicU32, Ordering::SeqCst},
+        time::Duration,
+    };
+
+    use matrix_sdk_base::{
+        event_cache_store::{EventCacheStore, EventCacheStoreError},
+        event_cache_store_integration_tests,
+        media::{MediaFormat, MediaRequest, MediaThumbnailSettings},
+    };
+    use matrix_sdk_test::async_test;
+    use once_cell::sync::Lazy;
+    use ruma::{events::room::MediaSource, media::Method, mxc_uri, uint};
+    use tempfile::{tempdir, TempDir};
+
+    use super::SqliteEventCacheStore;
+    use crate::utils::SqliteObjectExt;
+
+    static TMP_DIR: Lazy<TempDir> = Lazy::new(|| tempdir().unwrap());
+    static NUM: AtomicU32 = AtomicU32::new(0);
+
+    async fn get_event_cache_store() -> Result<SqliteEventCacheStore, EventCacheStoreError> {
+        let name = NUM.fetch_add(1, SeqCst).to_string();
+        let tmpdir_path = TMP_DIR.path().join(name);
+
+        tracing::info!("using event cache store @ {}", tmpdir_path.to_str().unwrap());
+
+        Ok(SqliteEventCacheStore::open(tmpdir_path.to_str().unwrap(), None).await.unwrap())
```

Comment:
It seems sqlite supports in-memory backends too, and our code half supports it (the path is optional); is this something we could use in testing contexts, instead of having to create temporary directories and counting the number of temporaries we've created?

Reply:
Where do you see the path as optional in our code?

We could use in-memory stores, but currently we rely heavily on `deadpool-sqlite` that requires a file. It's fine if we want to use in-memory store but we have to rewrite this to be able to interact with both `deadpool-sqlite` and `rusqlite` directly. In that case the layers we currently have (or something similar) might make more sense.

Reply:
> Where do you see the path as optional in our code?

`path` in the `SqliteEventCacheStore` impl; but it's probably a copy-pasto from the state store, which… never fills this path, and only uses it for debug purposes :facepalm: